File: bindings/scala/src/main/scala/proven/SafeUuid.scala
// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2025 Hyperpolymath

package com.hyperpolymath.proven

import java.security.SecureRandom
import java.util.UUID as JavaUUID
import scala.util.Try

/**
 * UUID version as per RFC 4122.
 */
enum UuidVersion:
  case V1   // Time-based
  case V2   // DCE Security
  case V3   // Name-based (MD5)
  case V4   // Random
  case V5   // Name-based (SHA-1)
  case V6   // Reordered time-based (draft)
  case V7   // Unix Epoch time-based (draft)
  case V8   // Custom (draft)
  case Unknown

object UuidVersion:
  /**
   * Parse version from integer (0-15).
   */
  def fromInt(v: Int): UuidVersion = v match
    case 1 => V1
    case 2 => V2
    case 3 => V3
    case 4 => V4
    case 5 => V5
    case 6 => V6
    case 7 => V7
    case 8 => V8
    case _ => Unknown

/**
 * UUID variant as per RFC 4122.
 */
enum UuidVariant:
  case NcsReserved      // Reserved for NCS backward compatibility
  case Rfc4122          // The variant specified in RFC 4122
  case MicrosoftReserved // Reserved for Microsoft backward compatibility
  case FutureReserved   // Reserved for future definition
  case Unknown

object UuidVariant:
  /**
   * Parse variant from the variant bits.
   */
  def fromBits(variantByte: Byte): UuidVariant =
    val v = (variantByte & 0xFF) >> 4
    if (v & 0x8) == 0 then NcsReserved
    else if (v & 0xC) == 0x8 then Rfc4122
    else if (v & 0xE) == 0xC then MicrosoftReserved
    else if (v & 0xE) == 0xE then FutureReserved
    else Unknown

/**
 * Represents a validated UUID with parsed components.
 *
 * @param timeLow          The low field of the timestamp (32 bits)
 * @param timeMid          The middle field of the timestamp (16 bits)
 * @param timeHiAndVersion The high field of the timestamp multiplexed with version (16 bits)
 * @param clockSeqHiAndRes The high field of clock sequence multiplexed with variant (8 bits)
 * @param clockSeqLow      The low field of clock sequence (8 bits)
 * @param node             The spatially unique node identifier (48 bits)
 */
case class Uuid(
    timeLow: Long,
    timeMid: Int,
    timeHiAndVersion: Int,
    clockSeqHiAndRes: Int,
    clockSeqLow: Int,
    node: Long
):
  /**
   * Get the UUID version.
   */
  def version: UuidVersion =
    UuidVersion.fromInt((timeHiAndVersion >> 12) & 0x0F)

  /**
   * Get the UUID variant.
   */
  def variant: UuidVariant =
    UuidVariant.fromBits(clockSeqHiAndRes.toByte)

  /**
   * Convert to the standard hyphenated string format.
   */
  def format: String =
    f"${timeLow}%08x-${timeMid}%04x-${timeHiAndVersion}%04x-${clockSeqHiAndRes}%02x${clockSeqLow}%02x-${node}%012x"

  /**
   * Convert to uppercase hyphenated string format.
   */
  def formatUppercase: String = format.toUpperCase

  /**
   * Convert to compact format (no hyphens).
   */
  def formatCompact: String =
    f"${timeLow}%08x${timeMid}%04x${timeHiAndVersion}%04x${clockSeqHiAndRes}%02x${clockSeqLow}%02x${node}%012x"

  /**
   * Convert to URN format.
   */
  def formatUrn: String = s"urn:uuid:$format"

  /**
   * Convert to byte array (16 bytes, big-endian).
   */
  def toBytes: Array[Byte] =
    val bytes = new Array[Byte](16)
    bytes(0) = ((timeLow >> 24) & 0xFF).toByte
    bytes(1) = ((timeLow >> 16) & 0xFF).toByte
    bytes(2) = ((timeLow >> 8) & 0xFF).toByte
    bytes(3) = (timeLow & 0xFF).toByte
    bytes(4) = ((timeMid >> 8) & 0xFF).toByte
    bytes(5) = (timeMid & 0xFF).toByte
    bytes(6) = ((timeHiAndVersion >> 8) & 0xFF).toByte
    bytes(7) = (timeHiAndVersion & 0xFF).toByte
    bytes(8) = clockSeqHiAndRes.toByte
    bytes(9) = clockSeqLow.toByte
    bytes(10) = ((node >> 40) & 0xFF).toByte
    bytes(11) = ((node >> 32) & 0xFF).toByte
    bytes(12) = ((node >> 24) & 0xFF).toByte
    bytes(13) = ((node >> 16) & 0xFF).toByte
    bytes(14) = ((node >> 8) & 0xFF).toByte
    bytes(15) = (node & 0xFF).toByte
    bytes

  /**
   * Convert to java.util.UUID.
   */
  def toJavaUuid: JavaUUID =
    val msb = (timeLow << 32) | (timeMid.toLong << 16) | timeHiAndVersion.toLong
    val lsb = (clockSeqHiAndRes.toLong << 56) | (clockSeqLow.toLong << 48) | node
    new JavaUUID(msb, lsb)

  /**
   * Check if this is the nil UUID (all zeros).
   */
  def isNil: Boolean =
    timeLow == 0 && timeMid == 0 && timeHiAndVersion == 0 &&
      clockSeqHiAndRes == 0 && clockSeqLow == 0 && node == 0

  /**
   * Check if this is the max UUID (all ones).
   */
  def isMax: Boolean =
    timeLow == 0xFFFFFFFFL && timeMid == 0xFFFF && timeHiAndVersion == 0xFFFF &&
      clockSeqHiAndRes == 0xFF && clockSeqLow == 0xFF && node == 0xFFFFFFFFFFFFL

  override def toString: String = format

object Uuid:
  private val secureRandom = new SecureRandom()

  private val HexPattern = """^[0-9a-fA-F]+$""".r

  /** The nil UUID (all zeros). */
  val Nil: Uuid = Uuid(0, 0, 0, 0, 0, 0)

  /** The max UUID (all ones). */
  val Max: Uuid = Uuid(0xFFFFFFFFL, 0xFFFF, 0xFFFF, 0xFF, 0xFF, 0xFFFFFFFFFFFFL)

  /**
   * Parse a UUID string in various formats.
   * Supports: standard (8-4-4-4-12), compact (32 hex), URN (urn:uuid:...), braced ({...}).
   */
  def parse(input: String): Option[Uuid] =
    val trimmed = input.trim

    // Handle URN format
    val withoutUrn = if trimmed.toLowerCase.startsWith("urn:uuid:") then
      trimmed.substring(9)
    else
      trimmed

    // Handle braced format
    val withoutBraces = if withoutUrn.startsWith("{") && withoutUrn.endsWith("}") then
      withoutUrn.substring(1, withoutUrn.length - 1)
    else
      withoutUrn

    // Remove hyphens
    val hexOnly = withoutBraces.replace("-", "")

    // Validate length and characters
    if hexOnly.length != 32 then return None
    if HexPattern.findFirstIn(hexOnly).isEmpty then return None

    Try {
      val timeLow = java.lang.Long.parseUnsignedLong(hexOnly.substring(0, 8), 16)
      val timeMid = Integer.parseUnsignedInt(hexOnly.substring(8, 12), 16)
      val timeHiAndVersion = Integer.parseUnsignedInt(hexOnly.substring(12, 16), 16)
      val clockSeqHiAndRes = Integer.parseUnsignedInt(hexOnly.substring(16, 18), 16)
      val clockSeqLow = Integer.parseUnsignedInt(hexOnly.substring(18, 20), 16)
      val node = java.lang.Long.parseUnsignedLong(hexOnly.substring(20, 32), 16)

      Uuid(timeLow, timeMid, timeHiAndVersion, clockSeqHiAndRes, clockSeqLow, node)
    }.toOption

  /**
   * Create from a byte array (16 bytes, big-endian).
   */
  def fromBytes(bytes: Array[Byte]): Option[Uuid] =
    if bytes.length != 16 then return None

    val timeLow = ((bytes(0) & 0xFFL) << 24) | ((bytes(1) & 0xFFL) << 16) |
      ((bytes(2) & 0xFFL) << 8) | (bytes(3) & 0xFFL)
    val timeMid = ((bytes(4) & 0xFF) << 8) | (bytes(5) & 0xFF)
    val timeHiAndVersion = ((bytes(6) & 0xFF) << 8) | (bytes(7) & 0xFF)
    val clockSeqHiAndRes = bytes(8) & 0xFF
    val clockSeqLow = bytes(9) & 0xFF
    val node = ((bytes(10) & 0xFFL) << 40) | ((bytes(11) & 0xFFL) << 32) |
      ((bytes(12) & 0xFFL) << 24) | ((bytes(13) & 0xFFL) << 16) |
      ((bytes(14) & 0xFFL) << 8) | (bytes(15) & 0xFFL)

    Some(Uuid(timeLow, timeMid, timeHiAndVersion, clockSeqHiAndRes, clockSeqLow, node))

  /**
   * Create from java.util.UUID.
   */
  def fromJavaUuid(uuid: JavaUUID): Uuid =
    val msb = uuid.getMostSignificantBits
    val lsb = uuid.getLeastSignificantBits

    val timeLow = (msb >> 32) & 0xFFFFFFFFL
    val timeMid = ((msb >> 16) & 0xFFFF).toInt
    val timeHiAndVersion = (msb & 0xFFFF).toInt
    val clockSeqHiAndRes = ((lsb >> 56) & 0xFF).toInt
    val clockSeqLow = ((lsb >> 48) & 0xFF).toInt
    val node = lsb & 0xFFFFFFFFFFFFL

    Uuid(timeLow, timeMid, timeHiAndVersion, clockSeqHiAndRes, clockSeqLow, node)

  /**
   * Generate a random UUID (version 4).
   */
  def v4(): Uuid =
    val bytes = new Array[Byte](16)
    secureRandom.nextBytes(bytes)

    // Set version to 4
    bytes(6) = ((bytes(6) & 0x0F) | 0x40).toByte

    // Set variant to RFC 4122
    bytes(8) = ((bytes(8) & 0x3F) | 0x80).toByte

    fromBytes(bytes).get

  /**
   * Generate a random UUID (version 4) - alias for v4().
   */
  def random(): Uuid = v4()

/**
 * Safe UUID operations with validation and parsing.
 */
object SafeUuid:

  /**
   * Check if string is a valid UUID.
   */
  def isValid(input: String): Boolean = Uuid.parse(input).isDefined

  /**
   * Parse a UUID string.
   */
  def parse(input: String): Option[Uuid] = Uuid.parse(input)

  /**
   * Generate a new random UUID (version 4).
   */
  def generate(): Uuid = Uuid.v4()

  /**
   * Generate a new random UUID as a string.
   */
  def generateString(): String = Uuid.v4().format

  /**
   * Format a UUID to the standard hyphenated lowercase format.
   */
  def format(uuid: Uuid): String = uuid.format

  /**
   * Format a UUID to uppercase.
   */
  def formatUppercase(uuid: Uuid): String = uuid.formatUppercase

  /**
   * Format a UUID to compact format (no hyphens).
   */
  def formatCompact(uuid: Uuid): String = uuid.formatCompact

  /**
   * Format a UUID to URN format.
   */
  def formatUrn(uuid: Uuid): String = uuid.formatUrn

  /**
   * Get the nil UUID.
   */
  def nil: Uuid = Uuid.Nil

  /**
   * Get the max UUID.
   */
  def max: Uuid = Uuid.Max

  /**
   * Check if a string represents the nil UUID.
   */
  def isNil(input: String): Boolean =
    Uuid.parse(input).exists(_.isNil)

  /**
   * Check if a string represents the max UUID.
   */
  def isMax(input: String): Boolean =
    Uuid.parse(input).exists(_.isMax)

  /**
   * Get the version of a UUID string.
   */
  def getVersion(input: String): Option[UuidVersion] =
    Uuid.parse(input).map(_.version)

  /**
   * Get the variant of a UUID string.
   */
  def getVariant(input: String): Option[UuidVariant] =
    Uuid.parse(input).map(_.variant)

  /**
   * Compare two UUID strings for equality (case-insensitive, format-agnostic).
   */
  def areEqual(a: String, b: String): Boolean =
    (Uuid.parse(a), Uuid.parse(b)) match
      case (Some(uuidA), Some(uuidB)) => uuidA == uuidB
      case _ => false

  /**
   * Normalize a UUID string to lowercase hyphenated format.
   */
  def normalize(input: String): Option[String] =
    Uuid.parse(input).map(_.format)

  /**
   * Convert UUID string to byte array.
   */
  def toBytes(input: String): Option[Array[Byte]] =
    Uuid.parse(input).map(_.toBytes)

  /**
   * Create UUID from byte array.
   */
  def fromBytes(bytes: Array[Byte]): Option[Uuid] =
    Uuid.fromBytes(bytes)

File: bindings/lua/proven/safe_uuid.lua
-- SPDX-License-Identifier: PMPL-1.0-or-later
-- SPDX-FileCopyrightText: 2025 Hyperpolymath

--- Safe UUID generation and validation following RFC 4122.
--- @module proven.safe_uuid

local safe_uuid = {}

--- UUID version types.
safe_uuid.Version = {
    V1 = 1,   -- Time-based
    V2 = 2,   -- DCE Security
    V3 = 3,   -- Name-based (MD5)
    V4 = 4,   -- Random
    V5 = 5,   -- Name-based (SHA-1)
    NIL = 0,  -- Nil UUID
}

--- UUID variant types.
safe_uuid.Variant = {
    NCS = "ncs",
    RFC4122 = "rfc4122",
    MICROSOFT = "microsoft",
    FUTURE = "future",
}

--- Namespace UUIDs (per RFC 4122).
safe_uuid.NAMESPACE_DNS = {
    0x6b, 0xa7, 0xb8, 0x10, 0x9d, 0xad, 0x11, 0xd1,
    0x80, 0xb4, 0x00, 0xc0, 0x4f, 0xd4, 0x30, 0xc8,
}

safe_uuid.NAMESPACE_URL = {
    0x6b, 0xa7, 0xb8, 0x11, 0x9d, 0xad, 0x11, 0xd1,
    0x80, 0xb4, 0x00, 0xc0, 0x4f, 0xd4, 0x30, 0xc8,
}

--- NIL UUID (all zeros).
safe_uuid.NIL = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}

-- UUID metatable for OOP behavior.
local Uuid = {}
Uuid.__index = Uuid

--- Create a new UUID object from bytes.
--- @param bytes table Array of 16 bytes
--- @return table UUID object
function Uuid.new(bytes)
    if type(bytes) ~= "table" or #bytes ~= 16 then
        return nil, "UUID requires exactly 16 bytes"
    end
    local self = setmetatable({}, Uuid)
    self._bytes = {}
    for i = 1, 16 do
        local byte_value = bytes[i]
        if type(byte_value) ~= "number" or byte_value < 0 or byte_value > 255 then
            return nil, "Invalid byte value at index " .. i
        end
        self._bytes[i] = math.floor(byte_value)
    end
    return self
end

--- Get the UUID bytes.
--- @return table Array of 16 bytes
function Uuid:bytes()
    local copy = {}
    for i = 1, 16 do
        copy[i] = self._bytes[i]
    end
    return copy
end

--- Get the UUID version.
--- @return number Version number (0-5)
function Uuid:version()
    local version_nibble = math.floor(self._bytes[7] / 16) % 16
    if version_nibble >= 1 and version_nibble <= 5 then
        return version_nibble
    end
    return safe_uuid.Version.NIL
end

--- Get the UUID variant.
--- @return string Variant identifier
function Uuid:variant()
    local byte8 = self._bytes[9]
    local high_bit = math.floor(byte8 / 128)
    if high_bit == 0 then
        return safe_uuid.Variant.NCS
    end
    local high_two_bits = math.floor(byte8 / 64)
    if high_two_bits == 2 then
        return safe_uuid.Variant.RFC4122
    end
    local high_three_bits = math.floor(byte8 / 32)
    if high_three_bits == 6 then
        return safe_uuid.Variant.MICROSOFT
    end
    return safe_uuid.Variant.FUTURE
end

--- Check if this is the nil UUID.
--- @return boolean
function Uuid:is_nil()
    for i = 1, 16 do
        if self._bytes[i] ~= 0 then
            return false
        end
    end
    return true
end

--- Format as canonical string (8-4-4-4-12).
--- @return string
function Uuid:format()
    local hex_chars = {}
    for i = 1, 16 do
        hex_chars[i] = string.format("%02x", self._bytes[i])
    end
    return table.concat(hex_chars, "", 1, 4) .. "-" ..
           table.concat(hex_chars, "", 5, 6) .. "-" ..
           table.concat(hex_chars, "", 7, 8) .. "-" ..
           table.concat(hex_chars, "", 9, 10) .. "-" ..
           table.concat(hex_chars, "", 11, 16)
end

--- Format as URN.
--- @return string
function Uuid:to_urn()
    return "urn:uuid:" .. self:format()
end

--- String representation.
function Uuid:__tostring()
    return self:format()
end

--- Equality comparison.
function Uuid:__eq(other)
    if getmetatable(other) ~= Uuid then
        return false
    end
    for i = 1, 16 do
        if self._bytes[i] ~= other._bytes[i] then
            return false
        end
    end
    return true
end

-- Expose the Uuid class
safe_uuid.Uuid = Uuid

--- Parse UUID from canonical string format.
--- @param uuid_string string The UUID string (8-4-4-4-12 format)
--- @return table|nil UUID object or nil on error
--- @return string|nil Error message if parsing failed
function safe_uuid.parse(uuid_string)
    if type(uuid_string) ~= "string" then
        return nil, "UUID must be a string"
    end

    if #uuid_string ~= 36 then
        return nil, "UUID must be 36 characters, got " .. #uuid_string
    end

    -- Validate hyphen positions
    if uuid_string:sub(9, 9) ~= "-" or
       uuid_string:sub(14, 14) ~= "-" or
       uuid_string:sub(19, 19) ~= "-" or
       uuid_string:sub(24, 24) ~= "-" then
        return nil, "Invalid UUID format: hyphens must be at positions 9, 14, 19, 24"
    end

    -- Remove hyphens and validate hex
    local hex_string = uuid_string:gsub("-", "")
    if #hex_string ~= 32 then
        return nil, "Invalid UUID length after removing hyphens"
    end

    -- Validate all characters are hex
    if not hex_string:match("^[0-9a-fA-F]+$") then
        return nil, "Invalid hex character in UUID"
    end

    -- Convert to bytes
    local bytes = {}
    for i = 1, 16 do
        local byte_hex = hex_string:sub(i * 2 - 1, i * 2)
        bytes[i] = tonumber(byte_hex, 16)
    end

    return Uuid.new(bytes)
end

--- Generate a v4 (random) UUID from provided random bytes.
--- @param random_bytes table Array of 16 random bytes
--- @return table|nil UUID object or nil on error
--- @return string|nil Error message if creation failed
function safe_uuid.v4_from_bytes(random_bytes)
    if type(random_bytes) ~= "table" or #random_bytes ~= 16 then
        return nil, "v4_from_bytes requires exactly 16 random bytes"
    end

    local bytes = {}
    for i = 1, 16 do
        bytes[i] = random_bytes[i]
    end

    -- Set version to 4 (bits 4-7 of byte 7 = 0100)
    local byte7_high = math.floor(bytes[7] / 16) % 16
    local byte7_low = bytes[7] % 16
    bytes[7] = 0x40 + byte7_low  -- 0100 in high nibble

    -- Set variant to RFC 4122 (bits 6-7 of byte 9 = 10)
    local byte9_masked = bytes[9] % 64  -- Clear top 2 bits
    bytes[9] = 0x80 + byte9_masked  -- 10 in high 2 bits

    return Uuid.new(bytes)
end

--- Check if string is valid UUID format.
--- @param uuid_string string The string to validate
--- @return boolean
function safe_uuid.is_valid(uuid_string)
    local uuid, _ = safe_uuid.parse(uuid_string)
    return uuid ~= nil
end

--- Create a nil UUID.
--- @return table UUID object
function safe_uuid.nil_uuid()
    return Uuid.new(safe_uuid.NIL)
end

--- Create UUID from bytes.
--- @param bytes table Array of 16 bytes
--- @return table|nil UUID object or nil on error
--- @return string|nil Error message if creation failed
function safe_uuid.from_bytes(bytes)
    return Uuid.new(bytes)
end

return safe_uuid
